ops/jenkins/bfd-build-platinum/Jenkinsfile

```diff
@@ -46,56 +120,76 @@ spec:
             steps {
                 script {
-                    awsAuth.assumeRole()
-
-                    if (platinumAmiId.length() != 0) {
-                        // Use latest images if sourceAmiId tag is set to the latest gold AMI ID
-                        echo "Using latest platinum AMI image: ${platinumAmiId}"
-                    } else if (platinumAmiId.length() == 0) {
-                        // Build base platinum if latest Platinum AMI has no sourceAmiId tag/sourceAmiId != gold AMI ID
-                        echo 'Deploying and tagging new platinum AMI'
-
-                        // Retrieve latest platium and assign the sourceAmiId tag value with the latest gold AMI ID.
-                        withCredentials([file(credentialsId: 'bfd-vault-password', variable: 'vaultPasswordFile')]) {
-                            // packer is always run from $repoRoot/ops/ansible/playbooks-ccs
-                            dir('ops/ansible/playbooks-ccs') {
-                                env.goldAmi = goldAmiId
-                                sh '''
-                                packer build -color=false -var vault_password_file="$vaultPasswordFile" \
-                                -var source_ami="${goldAmi}" \
-                                -var subnet_id=subnet-092c2a68bd18b34d1 \
-                                ../../packer/build_bfd-platinum.json
-                                '''
+                    currentStage = env.STAGE_NAME
+                    try {
+                        awsAuth.assumeRole()
+                        if (!params.FORCE_PLATINUM) {
+                            if (platinumAmiId.length() != 0) {
+                                // Use latest images if sourceAmiId tag is set to the latest gold AMI ID
+                                echo "Using latest platinum AMI image: ${platinumAmiId}"
+
+                                // Set current stage that the build succeeded
+                                sendNotifications('SUCCESS LATEST IMAGE', currentStage)
+                            } else if (platinumAmiId.length() == 0) {
+                                // Build base platinum if latest Platinum AMI has no sourceAmiId tag/sourceAmiId != gold AMI ID
+                                echo 'Deploying and tagging new platinum AMI'
+                                buildPlatinum(platinumAmiId, goldAmiId)
+
+                                // Set current stage that the build succeeded
+                                sendNotifications('SUCCESS NEW PLATINUM', currentStage)
                             }
+                        } else {
+                            echo 'Force deploying and tagging new platinum AMI'
+                            buildPlatinum(platinumAmiId, goldAmiId)
+
+                            // Set current stage that the build succeeded
+                            sendNotifications('SUCCESS NEW PLATINUM', currentStage)
                         }
```

Comment:
I don't think these conditions are that difficult to reason about. Then again, I don't think we should need to reason about them much at all. I wonder if it would be any clear if conditions tested the positives rather than the negatives here, e.g. `if (params.FORCE_PLATINUM) {...} else {...}` and `if (platinumAmIId.length() == 0) {...} else {...}`. What do you think?

Reply:
Logic is positive now